### ISSUE TITLE
Media aspect ratio not preserved

### ISSUE THREAD

**Issue by filipjakov** [hydra-synth/hydra-synth#120]
When using either the video or audio inputs from the examples:

```js
s0.initVideo("https://media.giphy.com/media/26ufplp8yheSKUE00/giphy.mp4", { flipY: true })
src(s0).out()
...
s0.initImage("https://upload.wikimedia.org/wikipedia/commons/2/25/Hydra-Foto.jpg")
src(s0).out()
```

the image or video gets fullscreened and the aspect ratio is not preserved, making them skewed.
Is there a parameter that can be used here that I am missing?
Is this something that I can add to the source? (For example: horizontal or vertical padds that would keep the aspect ratio)

**Comment by ritchse** [comment]
Hello ! I think most of us consider this to be a feature. It also has to do with how WebGL works. 
But I see that this can be frustrating, it would be nice to have some easy-to-change setting for this.

As of now I can share this piece of code that might help you for now:
```js
s0.initImage("https://upload.wikimedia.org/wikipedia/commons/2/25/Hydra-Foto.jpg")
src2 = (source) => src(source).scale(1,()=>(source.src.width/source.src.height)*(innerHeight/innerWidth))
src2(s0)
	.out()
```

**Comment by filipjakov** [comment]
@ritchse thank you, this is an amazing solution!

Is there also a quick way to make the image non-repeatable? Maybe even centered?

**Comment by ojack** [comment]
You can also edit the glsl https://hydra.ojack.xyz/?sketch_id=EfXLXaBxmiIsG1kY

```javascript

setFunction({
  name: 'srcNoRepeat',
  type: 'src',
  inputs: [
    {
      type: 'sampler2D',
      name: 'tex',
      default: NaN,
    }
  ],
  glsl:
`  vec4 c = texture2D(tex, _st);
   vec2 bl = step(vec2(0.0001),_st);       // bottom-left
   vec2 tr = step(vec2(0.0001),1.0-_st); // top right
   c.a = bl.x * bl.y * tr.x * tr.y;
   return c;`
})

s0.initImage("https://upload.wikimedia.org/wikipedia/commons/2/25/Hydra-Foto.jpg")
src2 = (source) => srcNoRepeat(source).scale(0.6,()=>(source.src.width/source.src.height)*(innerHeight/innerWidth))
src2(s0)
	.out()

```

 I actually wonder whether this should be the default behaviour for 'src'.....This function turns all parts of the canvas outside of the image transparent, and allows scaling without repetiton. . it is still possible to have repetitions using `repeat` and 'scrollX'.
 
 cc @micuat @ritchse

**Comment by ojack** [comment]
@filipjakov a hydra-only way is to scale a square mask with the same parameters as you scale the image.... for example

```javascript
s0.initImage("https://upload.wikimedia.org/wikipedia/commons/2/25/Hydra-Foto.jpg")

scaling = 0.6
aspect = (source) => source.src.width/source.src.height * innerHeight/innerWidth
src2 = (source) => src(source)
  .scale(scaling, aspect(source))
  .mask(shape(4, 0.999, 0).scale(scaling, aspect(source)))

src2(s0)
	.out()
```

**Comment by micuat** [comment]
> I actually wonder whether this should be the default behaviour for 'src'

I feel this will "break" some sketches so we should not change the behavior until we have a "versioning" system that users can select which version to use.

Another related problem is that a webcam source doesn't have resolution info if I remember correctly (I had to open media stream by JavaScript and then use `s0.init()` to get dimensions)

**Comment by micuat** [comment]
and for the non-wrapping scaling, here's a bit more concise version (+works on load as bonus)

```javascript
s0.initImage("https://upload.wikimedia.org/wikipedia/commons/2/25/Hydra-Foto.jpg")

scaledSrc = (source, scaleAmount = 1) => src(source)
	.mask(shape(4, 0.999, 0))
		.scale(scaleAmount, ()=>source.src?.width / source.src?.height * innerHeight / innerWidth)

scaledSrc(s0, 0.4)
	.out()
```

**Comment by ritchse** [comment]
I agree with @micuat . Also I kinda like the current behavior as default. But it'd be nice to have an option.
This is another one of those optional things that we'll have to come up with some way of letting users change easily. Such as wrapping, framebuffers' interpolation mode, alpha mode, etc.
Maybe we should open an issue about that? Or maybe a hackmd document where we can sketch some ideas about how we'd like to see it implemented (from the user's perspective)

**Comment by filipjakov** [comment]
Thanks everyone, this was really helpful!

**Comment by filipjakov** [comment]
@micuat 
> Another related problem is that a webcam source doesn't have resolution info if I remember correctly (I had to open media stream by JavaScript and then use s0.init() to get dimensions)

I tried using [videoWidth](https://developer.mozilla.org/en-US/docs/Web/API/HTMLVideoElement/videoWidth)/[videoheight](https://developer.mozilla.org/en-US/docs/Web/API/HTMLVideoElement/videoHeight) to get the intrinsic dimensions of the camera stream, seems to be working:

```js
s0.initCam()

console.log(s0.src.videoWidth, s0.src.videoHeight)

scaledSrc = (source, scaleAmount = 1) => src(source)
	.mask(shape(4, 0.999, 0))
		.scale(scaleAmount, ()=>source.src?.videoWidth / source.src?.videoHeight * innerHeight / innerWidth)

scaledSrc(s0, 1)
	.out()
```

**Comment by ojack** [comment]
> > I actually wonder whether this should be the default behaviour for 'src'
> 
> I feel this will "break" some sketches so we should not change the behavior until we have a "versioning" system that users can select which version to use.
> 
> Another related problem is that a webcam source doesn't have resolution info if I remember correctly (I had to open media stream by JavaScript and then use `s0.init()` to get dimensions)


definitely agree that this shouldn't happen without a versioning system! But I mean once we *do* have a versioning system.
